File: frontend/src/services/eventService.ts
/**
 * Event Service
 *
 * API client for Event operations.
 */

import type {
  Event,
  EventCreateRequest,
  EventUpdateRequest,
  EventStatusUpdateRequest,
  EventPostponeRequest,
  EventRSVPUpdateRequest,
  EventListResponse,
  EventStatistics,
  EventListParams,
  UpcomingEventsParams,
  DateRangeEventsParams,
  EventsByTypeParams,
  StatisticsParams
} from '@/types/event'

const API_BASE = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000'

class EventService {
  /**
   * Create a new event
   */
  async createEvent(
    data: EventCreateRequest,
    createdById: string
  ): Promise<Event> {
    const response = await fetch(
      `${API_BASE}/api/v1/events?created_by_id=${createdById}`,
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(data)
      }
    )

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.detail || 'Failed to create event')
    }

    return response.json()
  }

  /**
   * Get all events with filters and pagination
   */
  async getEvents(params: EventListParams): Promise<EventListResponse> {
    const queryParams = new URLSearchParams()
    queryParams.append('school_id', params.school_id)

    if (params.start_date) queryParams.append('start_date', params.start_date)
    if (params.end_date) queryParams.append('end_date', params.end_date)
    if (params.event_type) queryParams.append('event_type', params.event_type)
    if (params.status) queryParams.append('status', params.status)
    if (params.page) queryParams.append('page', params.page.toString())
    if (params.limit) queryParams.append('limit', params.limit.toString())

    const response = await fetch(
      `${API_BASE}/api/v1/events?${queryParams.toString()}`
    )

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.detail || 'Failed to fetch events')
    }

    return response.json()
  }

  /**
   * Get upcoming events
   */
  async getUpcomingEvents(params: UpcomingEventsParams): Promise<Event[]> {
    const queryParams = new URLSearchParams()
    queryParams.append('school_id', params.school_id)

    if (params.days_ahead)
      queryParams.append('days_ahead', params.days_ahead.toString())
    if (params.limit) queryParams.append('limit', params.limit.toString())

    const response = await fetch(
      `${API_BASE}/api/v1/events/upcoming?${queryParams.toString()}`
    )

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.detail || 'Failed to fetch upcoming events')
    }

    return response.json()
  }

  /**
   * Get events by date range
   */
  async getEventsByDateRange(
    params: DateRangeEventsParams
  ): Promise<Event[]> {
    const queryParams = new URLSearchParams()
    queryParams.append('school_id', params.school_id)
    queryParams.append('start_date', params.start_date)
    queryParams.append('end_date', params.end_date)

    const response = await fetch(
      `${API_BASE}/api/v1/events/date-range?${queryParams.toString()}`
    )

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.detail || 'Failed to fetch events by date range')
    }

    return response.json()
  }

  /**
   * Get events by type
   */
  async getEventsByType(params: EventsByTypeParams): Promise<EventListResponse> {
    const queryParams = new URLSearchParams()
    queryParams.append('school_id', params.school_id)

    if (params.page) queryParams.append('page', params.page.toString())
    if (params.limit) queryParams.append('limit', params.limit.toString())

    const response = await fetch(
      `${API_BASE}/api/v1/events/type/${params.event_type}?${queryParams.toString()}`
    )

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.detail || 'Failed to fetch events by type')
    }

    return response.json()
  }

  /**
   * Get events by organizer
   */
  async getEventsByOrganizer(
    organizerId: string,
    page: number = 1,
    limit: number = 50
  ): Promise<EventListResponse> {
    const queryParams = new URLSearchParams()
    queryParams.append('page', page.toString())
    queryParams.append('limit', limit.toString())

    const response = await fetch(
      `${API_BASE}/api/v1/events/organizer/${organizerId}?${queryParams.toString()}`
    )

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.detail || 'Failed to fetch events by organizer')
    }

    return response.json()
  }

  /**
   * Get events by room
   */
  async getEventsByRoom(
    roomId: string,
    startDate?: string,
    endDate?: string
  ): Promise<Event[]> {
    const queryParams = new URLSearchParams()
    if (startDate) queryParams.append('start_date', startDate)
    if (endDate) queryParams.append('end_date', endDate)

    const response = await fetch(
      `${API_BASE}/api/v1/events/room/${roomId}?${queryParams.toString()}`
    )

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.detail || 'Failed to fetch events by room')
    }

    return response.json()
  }

  /**
   * Get RSVP events
   */
  async getRSVPEvents(
    schoolId: string,
    upcomingOnly: boolean = true
  ): Promise<Event[]> {
    const queryParams = new URLSearchParams()
    queryParams.append('school_id', schoolId)
    queryParams.append('upcoming_only', upcomingOnly.toString())

    const response = await fetch(
      `${API_BASE}/api/v1/events/rsvp?${queryParams.toString()}`
    )

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.detail || 'Failed to fetch RSVP events')
    }

    return response.json()
  }

  /**
   * Get recurring events
   */
  async getRecurringEvents(schoolId: string): Promise<Event[]> {
    const response = await fetch(
      `${API_BASE}/api/v1/events/recurring?school_id=${schoolId}`
    )

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.detail || 'Failed to fetch recurring events')
    }

    return response.json()
  }

  /**
   * Get event statistics
   */
  async getStatistics(params: StatisticsParams): Promise<EventStatistics> {
    const queryParams = new URLSearchParams()
    queryParams.append('school_id', params.school_id)

    if (params.start_date) queryParams.append('start_date', params.start_date)
    if (params.end_date) queryParams.append('end_date', params.end_date)

    const response = await fetch(
      `${API_BASE}/api/v1/events/statistics/summary?${queryParams.toString()}`
    )

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.detail || 'Failed to fetch event statistics')
    }

    return response.json()
  }

  /**
   * Get event by ID
   */
  async getEventById(id: string): Promise<Event> {
    const response = await fetch(`${API_BASE}/api/v1/events/${id}`)

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.detail || 'Failed to fetch event')
    }

    return response.json()
  }

  /**
   * Update an event
   */
  async updateEvent(
    id: string,
    data: EventUpdateRequest,
    updatedById: string
  ): Promise<Event> {
    const response = await fetch(
      `${API_BASE}/api/v1/events/${id}?updated_by_id=${updatedById}`,
      {
        method: 'PUT',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(data)
      }
    )

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.detail || 'Failed to update event')
    }

    return response.json()
  }

  /**
   * Update event status
   */
  async updateEventStatus(
    id: string,
    data: EventStatusUpdateRequest,
    updatedById: string
  ): Promise<Event> {
    const response = await fetch(
      `${API_BASE}/api/v1/events/${id}/status?updated_by_id=${updatedById}`,
      {
        method: 'PATCH',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(data)
      }
    )

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.detail || 'Failed to update event status')
    }

    return response.json()
  }

  /**
   * Postpone an event
   */
  async postponeEvent(
    id: string,
    data: EventPostponeRequest,
    updatedById: string
  ): Promise<Event> {
    const response = await fetch(
      `${API_BASE}/api/v1/events/${id}/postpone?updated_by_id=${updatedById}`,
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(data)
      }
    )

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.detail || 'Failed to postpone event')
    }

    return response.json()
  }

  /**
   * Increment RSVP count
   */
  async incrementRSVP(
    id: string,
    data: EventRSVPUpdateRequest
  ): Promise<Event> {
    const response = await fetch(
      `${API_BASE}/api/v1/events/${id}/rsvp/increment`,
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(data)
      }
    )

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.detail || 'Failed to increment RSVP count')
    }

    return response.json()
  }

  /**
   * Decrement RSVP count
   */
  async decrementRSVP(
    id: string,
    data: EventRSVPUpdateRequest
  ): Promise<Event> {
    const response = await fetch(
      `${API_BASE}/api/v1/events/${id}/rsvp/decrement`,
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(data)
      }
    )

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.detail || 'Failed to decrement RSVP count')
    }

    return response.json()
  }

  /**
   * Delete an event
   */
  async deleteEvent(id: string, deletedById: string): Promise<void> {
    const response = await fetch(
      `${API_BASE}/api/v1/events/${id}?deleted_by_id=${deletedById}`,
      {
        method: 'DELETE'
      }
    )

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.detail || 'Failed to delete event')
    }
  }
}

export default new EventService()

File: backend/services/event_service.py
"""
Event Service

Business logic layer for Event operations.
"""

from typing import List, Optional, Dict, Any
from sqlalchemy.ext.asyncio import AsyncSession
from repositories.event_repository import EventRepository
from repositories.room_repository import RoomRepository
from repositories.user_repository import UserRepository
from models.event import Event
from datetime import date, time
import uuid


class EventService:
    """Service layer for Event business logic"""

    def __init__(self, session: AsyncSession):
        self.session = session
        self.repository = EventRepository(session)
        self.room_repository = RoomRepository(session)
        self.user_repository = UserRepository(session)

    async def create_event(
        self,
        school_id: uuid.UUID,
        title: str,
        event_type: str,
        start_date: date,
        end_date: date,
        created_by_id: uuid.UUID,
        description: Optional[str] = None,
        start_time: Optional[time] = None,
        end_time: Optional[time] = None,
        is_all_day: bool = False,
        location: Optional[str] = None,
        room_id: Optional[uuid.UUID] = None,
        target_audience: Optional[str] = None,
        grade_levels: Optional[List[int]] = None,
        class_ids: Optional[List[uuid.UUID]] = None,
        organizer_id: Optional[uuid.UUID] = None,
        organizer_name: Optional[str] = None,
        status: str = "scheduled",
        is_recurring: bool = False,
        recurrence_pattern: Optional[str] = None,
        recurrence_end_date: Optional[date] = None,
        requires_rsvp: bool = False,
        max_attendees: Optional[int] = None,
        color: Optional[str] = None,
        attachments: Optional[Dict] = None
    ) -> Event:
        """Create a new event"""

        # Validate dates
        if end_date < start_date:
            raise ValueError("End date must be on or after start date")

        # Validate room if provided
        if room_id:
            room = await self.room_repository.get_by_id(room_id)
            if not room:
                raise ValueError("Room not found")

            # Check for room conflicts
            conflict = await self.repository.check_room_conflict(
                room_id=room_id,
                start_date=start_date,
                end_date=end_date,
                start_time=start_time,
                end_time=end_time
            )
            if conflict:
                raise ValueError(f"Room is already booked for event '{conflict.title}' during this time")

        # Validate organizer if provided
        if organizer_id:
            organizer = await self.user_repository.get_by_id(organizer_id)
            if not organizer:
                raise ValueError("Organizer user not found")

        # Validate event type
        valid_types = ['assembly', 'exam', 'holiday', 'meeting', 'parent_conference',
                      'field_trip', 'sports', 'performance', 'workshop', 'other']
        if event_type not in valid_types:
            raise ValueError(f"Invalid event type. Must be one of: {', '.join(valid_types)}")

        # Validate status
        valid_statuses = ['scheduled', 'in_progress', 'completed', 'cancelled', 'postponed']
        if status not in valid_statuses:
            raise ValueError(f"Invalid status. Must be one of: {', '.join(valid_statuses)}")

        # Validate target audience
        if target_audience:
            valid_audiences = ['all_school', 'grade_level', 'class', 'custom']
            if target_audience not in valid_audiences:
                raise ValueError(f"Invalid target audience. Must be one of: {', '.join(valid_audiences)}")

        event_data = {
            'school_id': school_id,
            'title': title,
            'description': description,
            'event_type': event_type,
            'start_date': start_date,
            'end_date': end_date,
            'start_time': start_time,
            'end_time': end_time,
            'is_all_day': is_all_day,
            'location': location,
            'room_id': room_id,
            'target_audience': target_audience,
            'grade_levels': grade_levels,
            'class_ids': class_ids,
            'organizer_id': organizer_id,
            'organizer_name': organizer_name,
            'status': status,
            'is_recurring': is_recurring,
            'recurrence_pattern': recurrence_pattern,
            'recurrence_end_date': recurrence_end_date,
            'requires_rsvp': requires_rsvp,
            'max_attendees': max_attendees,
            'current_attendees': 0,
            'color': color,
            'reminder_sent': False,
            'attachments': attachments
        }

        return await self.repository.create(event_data, created_by_id)

    async def update_event(
        self,
        event_id: uuid.UUID,
        updated_by_id: uuid.UUID,
        **updates
    ) -> Optional[Event]:
        """Update an event"""
        event = await self.repository.get_by_id(event_id)
        if not event:
            return None

        # Validate dates if being updated
        start_date = updates.get('start_date', event.start_date)
        end_date = updates.get('end_date', event.end_date)

        if end_date < start_date:
            raise ValueError("End date must be on or after start date")

        # Check room conflicts if room or dates are being updated
        if 'room_id' in updates or 'start_date' in updates or 'end_date' in updates:
            room_id = updates.get('room_id', event.room_id)
            if room_id:
                start_time = updates.get('start_time', event.start_time)
                end_time = updates.get('end_time', event.end_time)

                conflict = await self.repository.check_room_conflict(
                    room_id=room_id,
                    start_date=start_date,
                    end_date=end_date,
                    start_time=start_time,
                    end_time=end_time,
                    exclude_event_id=event_id
                )
                if conflict:
                    raise ValueError(f"Room is already booked for event '{conflict.title}' during this time")

        return await self.repository.update(event_id, updates, updated_by_id)

    async def cancel_event(
        self,
        event_id: uuid.UUID,
        updated_by_id: uuid.UUID
    ) -> Optional[Event]:
        """Cancel an event"""
        return await self.repository.update(
            event_id,
            {'status': 'cancelled'},
            updated_by_id
        )

    async def postpone_event(
        self,
        event_id: uuid.UUID,
        new_start_date: date,
        new_end_date: date,
        updated_by_id: uuid.UUID
    ) -> Optional[Event]:
        """Postpone an event to new dates"""
        event = await self.repository.get_by_id(event_id)
        if not event:
            return None

        if new_end_date < new_start_date:
            raise ValueError("End date must be on or after start date")

        # Check room conflicts for new dates
        if event.room_id:
            conflict = await self.repository.check_room_conflict(
                room_id=event.room_id,
                start_date=new_start_date,
                end_date=new_end_date,
                start_time=event.start_time,
                end_time=event.end_time,
                exclude_event_id=event.id
            )
            if conflict:
                raise ValueError(f"Room is already booked for event '{conflict.title}' during the new dates")

        return await self.repository.update(
            event.id,
            {
                'start_date': new_start_date,
                'end_date': new_end_date,
                'status': 'postponed'
            },
            updated_by_id
        )

    async def increment_attendees(
        self,
        event_id: uuid.UUID,
        count: int = 1
    ) -> Optional[Event]:
        """Increment the attendee count for an RSVP event"""
        event = await self.repository.get_by_id(event_id)
        if not event:
            return None

        if not event.requires_rsvp:
            raise ValueError("This event does not require RSVP")

        if event.max_attendees and (event.current_attendees + count) > event.max_attendees:
            raise ValueError("Event is at capacity")

        new_count = event.current_attendees + count
        return await self.repository.update(
            event_id,
            {'current_attendees': new_count},
            None  # No user update needed for count increment
        )

    async def decrement_attendees(
        self,
        event_id: uuid.UUID,
        count: int = 1
    ) -> Optional[Event]:
        """Decrement the attendee count for an RSVP event"""
        event = await self.repository.get_by_id(event_id)
        if not event:
            return None

        if not event.requires_rsvp:
            raise ValueError("This event does not require RSVP")

        new_count = max(0, event.current_attendees - count)
        return await self.repository.update(
            event_id,
            {'current_attendees': new_count},
            None
        )

    async def get_upcoming_events(
        self,
        school_id: uuid.UUID,
        days_ahead: int = 30,
        limit: int = 10
    ) -> List[Event]:
        """Get upcoming events"""
        return await self.repository.get_upcoming_events(school_id, days_ahead, limit)

    async def get_events_by_date_range(
        self,
        school_id: uuid.UUID,
        start_date: date,
        end_date: date
    ) -> List[Event]:
        """Get events within a date range"""
        return await self.repository.get_by_date_range(school_id, start_date, end_date)

    async def get_events_by_type(
        self,
        school_id: uuid.UUID,
        event_type: str,
        page: int = 1,
        limit: int = 50
    ) -> tuple[List[Event], int]:
        """Get events by type"""
        return await self.repository.get_by_type(school_id, event_type, page, limit)

    async def delete_event(
        self,
        event_id: uuid.UUID,
        deleted_by_id: uuid.UUID
    ) -> bool:
        """Soft delete an event"""
        return await self.repository.delete(event_id, deleted_by_id)

    async def get_statistics(
        self,
        school_id: uuid.UUID,
        start_date: Optional[date] = None,
        end_date: Optional[date] = None
    ) -> Dict[str, Any]:
        """Get event statistics"""
        return await self.repository.get_statistics(school_id, start_date, end_date)
